ext/js/pages/settings/anki-deck-generator-controller.js

```diff
@@ -0,0 +1,588 @@
+        };
+        const template = this._getAnkiTemplate(options);
+        const deckOptionsFields = options.anki.terms.fields;
+        const {general: {resultOutputMode, glossaryLayoutMode, compactTags}} = options;
+        const fields = [];
+        for (const deckField in deckOptionsFields) {
+            if (Object.prototype.hasOwnProperty.call(deckOptionsFields, deckField)) {
+                fields.push([deckField, deckOptionsFields[deckField]]);
+            }
+        }
+        const idleTimeout = (Number.isFinite(options.anki.downloadTimeout) && options.anki.downloadTimeout > 0 ? options.anki.downloadTimeout : null);
+        const mediaOptions = addMedia ? {audio: {sources: options.audio.sources, preferredAudioIndex: null, idleTimeout: idleTimeout}} : null;
+        const requirements = addMedia ? [...this._getDictionaryEntryMedia(dictionaryEntry), {type: 'audio'}] : [];
+        const {note} = await this._ankiNoteBuilder.createNote(/** @type {import('anki-note-builder').CreateNoteDetails} */ ({
+            dictionaryEntry,
+            mode,
+            context,
+            template,
+            deckName: this._activeAnkiDeck,
+            modelName: this._activeNoteType,
+            fields: fields,
+            resultOutputMode,
+            glossaryLayoutMode,
+            compactTags,
+            tags: options.anki.tags,
+            mediaOptions: mediaOptions,
+            requirements: requirements,
+            duplicateScope: options.anki.duplicateScope,
+            duplicateScopeCheckAllModels: options.anki.duplicateScopeCheckAllModels
+        }));
+        return note;
+    }
+
+    /**
+     * @param {string} text
+     * @param {import('settings').OptionsContext} optionsContext
+     * @returns {Promise<?{dictionaryEntry: import('dictionary').TermDictionaryEntry, text: string}>}
+     */
+    async _getDictionaryEntry(text, optionsContext) {
+        const {dictionaryEntries} = await this._settingsController.application.api.termsFind(text, {}, optionsContext);
+        if (dictionaryEntries.length === 0) { return null; }
+
+        return {
+            dictionaryEntry: /** @type {import('dictionary').TermDictionaryEntry} */ (dictionaryEntries[0]),
+            text: text
+        };
+    }
+
+    /**
+     * @param {import('dictionary').TermDictionaryEntry} dictionaryEntry
+     * @returns {Array<object>}
+     */
+    _getDictionaryEntryMedia(dictionaryEntry) {
+        const media = [];
+        const definitions = dictionaryEntry.definitions;
+        for (const definition of definitions) {
+            const paths = this._findAllByKey(definition, 'path');
+            for (const path of paths) {
+                media.push({dictionary: definition.dictionary, path: path, type: 'dictionaryMedia'});
+            }
+        }
+        return media;
+    }
+
+    /**
+     * @param {object} obj
+     * @param {string} keyToFind
+     * @returns {Array<string>}
+     */
+    _findAllByKey(obj, keyToFind) {
```

Comment:
Find all what by key?

Reply:
It searches for any contents for matching keys. The contents of what is found could be anything but in this case I'm just using it for fetching path strings in deeply nested structured content objects and arrays.

Reply:
```suggestion
    _findAllValuesByKey(obj, keyToFind) {
``` 
or 
```suggestion
    _findPathsByKey(obj) {
```
since it's only used for finding paths (not seeing any other use cases for it). It also prevents other developers from using it without understanding the rough edges, one of which is this function potentially getting caught in a recursive loop:
```js
const obj = {'foo': 'bar'};
obj['baz'] = obj;

console.log(_findAllByKey(obj, 'foo'));
``` 
leads to `Maximum call stack size exceeded`

Reply:
Good point. Limited it to 'path's only.